src/training/scheduler.h

```diff
@@ -535,9 +535,14 @@ class Scheduler : public TrainingObserver {
 
   void actAfterEpoch(TrainingState& state) override {
     // stop if data streaming from STDIN is stopped for a TSV input
-    std::string firstPath = options_->get<std::vector<std::string>>("train-sets")[0];
-    if(options_->get<bool>("tsv", false) && (firstPath == "stdin" || firstPath == "-"))
-      endOfStdin_ = true;
+    if (options_->has("training-sets")) {
+      auto trainingSets = options_->get<std::vector<std::string>>("train-sets");
+      if (trainingSets.size() > 0) {
+        std::string firstPath = options_->get<std::vector<std::string>>("train-sets")[0];
+        if(options_->get<bool>("tsv", false) && (firstPath == "stdin" || firstPath == "-"))
+          endOfStdin_ = true;
+      }
+    }
```

Comment:
Was this change done for debugging purposes or is it now required with marian-adaptive?

Reply:
It's required. When running self-adaptive marian in server mode the `"training-sets"` option isn't present because the training sentences are passed in via the request body so the scheduler would crash when trying to read the option.